core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -683,7 +694,28 @@ final class ShuffleBlockFetcherIterator(
             }
           }
 
-        case FailureFetchResult(blockId, mapIndex, address, e) =>
+        // Catching OOM and do something based on it is only a workaround for handling the
+        // Netty OOM issue, which is not the best way towards memory management. We can
+        // get rid of it when we find a way to manage Netty's memory precisely.
+        case FailureFetchResult(blockId, mapIndex, address, size, isNetworkReqDone, e)
+            if e.isInstanceOf[OutOfDirectMemoryError] || e.isInstanceOf[NettyOutOfMemoryError] =>
+          assert(address != blockManager.blockManagerId &&
+            !hostLocalBlocks.contains(blockId -> mapIndex),
+            "Netty OOM error should only happen on remote fetch requests")
+          logWarning(s"Failed to fetch block $blockId due to Netty OOM, will retry", e)
+          NettyUtils.isNettyOOMOnShuffle = true
+          numBlocksInFlightPerAddress(address) = numBlocksInFlightPerAddress(address) - 1
+          bytesInFlight -= size
+          if (isNetworkReqDone) {
+            reqsInFlight -= 1
+            logDebug("Number of requests in flight " + reqsInFlight)
+          }
+          val defReqQueue =
+            deferredFetchRequests.getOrElseUpdate(address, new Queue[FetchRequest]())
+          defReqQueue.enqueue(FetchRequest(address, Array(FetchBlockInfo(blockId, size, mapIndex))))
```

Comment:
How many times will we retry ?

Reply:
It will keep retry until succeeds. So in the worst case, it will retry for the times of the number of total requests, which would fetch one by one. Theoretically, it should succeed in the end (unless that netty is even not able to serve one request at a time) and won't be an endless retry.

Reply:
That would be a lot of load on the remote shuffle service/remote executor right ?
Perhaps we should bound this somehow ?

Reply:
If an executor is not assigned enough off-heap memory. Is it possible in this case that it keeps retrying forever? Let's say there is some skew and one of the blocks is large and the executor isn't assigned enough memory. Whenever this block is fetched Netty OOMs. In this case will this keep retrying? Maybe I am missing something.

Reply:
> That would be a lot of load on the remote shuffle service/remote executor right ?
Perhaps we should bound this somehow ?

This's a good point! I think there's a tradeoff between the cost of the reload and the cost stage recomputation.

Reply:
> When an executor is not assigned enough off-heap memory, is it possible that in this case, it keeps retrying forever? Let's say there is some skew and one of the blocks is large and whenever this block is fetched Netty OOMs. In this case will this keep retrying? Maybe I am missing something.

To clarify first, Netty memory is not a part of executor memory,  but a separate memory region (logically) instead. Users could use `io.netty.maxDirectMemory` to set and Netty uses [maxDirectMemory0](https://github.com/netty/netty/blob/df53de5b687cd25ba1079b556318da2b773cf4f1/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L1140) by default if not configured manually.

In the case of skew/large block, Spark has `spark.network.maxRemoteBlockSizeFetchToMem` (200M by default) to save the block into disk instead of holding the whole block in Netty memory. 

And it's true that we can fall into endless retry if Netty memory is still less enough to serve a small block..

I can add a check like:

```
if (NettyUtils.maxDirectMemory() < blockSize) {
 // throw fetch failure directly instead of retrying
}
```

Does it look good to you? @mridulm @otterc

Reply:
> In a large-scale cluster, I think there should be already plenty of loads. So I'd prefer to avoid stage recomputation as possible as we can.

Agree on need to minimize stage recomputation if we can judiciously retry.
The concern is that repeated fetches can cause issues for all applications relying on the shuffle service - not just a single application; and so causes an impact across apps.

Reply:
+CC @otterc. Can you take a look at @Ngone51's idea above pls ? Thx

Reply:
Thanks @Ngone51 for the clarification. I have a better understanding of the netty memory now. So, I think your recent change below would work to avoid endless retries
```
              if PlatformDependent.maxDirectMemory() > ( 1.5 * size) => defer the request 
``` 
But I have a follow up question here which comes from the other change in this PR. For a remote response, seems like the buffer could be either a `NettyManagedBuffer` or not. When it is **not** a `NettyManagedBuffer` then can there still be a `OutOfDirectMemoryError`?

Reply:
> When it is not a NettyManagedBuffer then can there still be a OutOfDirectMemoryError?

It's possible.

In general, for a remote request, it must rely on Netty to receive the data sent from the server, which of course consumes Netty Memory.  So it doesn't matter whatever the final stored type is, e.g., `NettyManagedBuffer` or `FileSegmentManagedBuffer`.